diff --git a/lib/rack_warden.rb b/lib/rack_warden.rb
index a9cc87e..a75c930 100644
--- a/lib/rack_warden.rb
+++ b/lib/rack_warden.rb
@@ -6,7 +6,8 @@ module RackWarden
 end
 
 require "sinatra/base"
-require "sinatra/flash"
+#require "sinatra/flash"
+require "rack/flash"
 require 'bcrypt'
 require 'data_mapper'
 require 'warden'
diff --git a/lib/rack_warden/app.rb b/lib/rack_warden/app.rb
index 08e5915..7e861ca 100644
--- a/lib/rack_warden/app.rb
+++ b/lib/rack_warden/app.rb
@@ -4,7 +4,7 @@
 module RackWarden
   class App < Sinatra::Base
   
-    register Sinatra::Flash
+    #register Sinatra::Flash
         
     set :config_files, [ENV['RACK_WARDEN_CONFIG_FILE'], 'rack_warden.yml', 'config/rack_warden.yml'].compact.uniq
     set :layout, :'rw_layout.html'
@@ -72,6 +72,8 @@ module RackWarden
 	  
 	  # Main RackWarden::App class setup.
 	  def self.initialize_app
+	  	use Rack::Flash, :accessorize=>:rwarden
+	  
 	  	initialize_logging
 	  	logger.warn "RW initializing RackWarden::App in process #{$0}"
 	  	initialize_config_files
diff --git a/lib/rack_warden/routes.rb b/lib/rack_warden/routes.rb
index 4dfb5c5..51fb573 100644
--- a/lib/rack_warden/routes.rb
+++ b/lib/rack_warden/routes.rb
@@ -17,7 +17,7 @@ module RackWarden
 				  if User.count > 0
 				    erb :'rw_login.html', :layout=>settings.layout
 				  else
-				    flash(:rwarden)[:error] = warden.message || "Please create an admin account"
+				    flash.rwarden = {:error => (warden.message || "Please create an admin account")}
 				    redirect url('/auth/new', false)
 				  end
 				end
@@ -25,7 +25,7 @@ module RackWarden
 				post '/auth/login' do
 				  warden.authenticate!
 				
-				  flash(:rwarden)[:success] = warden.message || "Successful login"
+				  flash[:rwarden][:success] = warden.message || "Successful login"
 				
 				  return_to
 				end
@@ -33,7 +33,7 @@ module RackWarden
 				get '/auth/logout' do
 				  #warden.raw_session.inspect
 				  warden.logout
-				  flash(:rwarden)[:success] = 'You have been logged out'
+				  flash[:rwarden][:success] = 'You have been logged out'
 				  redirect url(settings.default_route, false)
 				end
 				
@@ -49,12 +49,12 @@ module RackWarden
 				  @user = User.new(params['user'])
 				  if @user.save
 				    warden.set_user(@user)
-				  	flash(:rwarden)[:success] = warden.message || "Account created"
+				  	flash.rwarden[:success] = (warden.message || "Account created")
 				  	App.logger.info "RW /auth/create succeeded for '#{user.username rescue nil}' #{@user.errors.entries}"
 				    #redirect session[:return_to] || url(settings.default_route, false)
 				    return_to
 				  else
-				  	flash(:rwarden)[:error] = "#{warden.message} => #{@user.errors.entries.join('. ')}"
+				  	flash[:rwarden][:error] = "#{warden.message} => #{@user.errors.entries.join('. ')}"
 				  	App.logger.info "RW /auth/create errors for '#{user.username rescue nil}' #{@user.errors.entries}"
 				  	redirect back #url('/auth/new', false)
 				  end
@@ -67,10 +67,10 @@ module RackWarden
 				  App.logger.debug "RW will return-to #{session[:return_to]}"
 				  App.logger.debug warden
 				  # if User.count > 0
-				    flash(:rwarden)[:error] = warden.message || "Please login to continue"
+				    flash[:rwarden][:error] = warden.message || "Please login to continue"
 				    redirect url('/auth/login', false)
 				  # else
-				  #   flash(:rwarden)[:error] = warden.message || "Please create an admin account"
+				  #   flash[:rwarden][:error] = warden.message || "Please create an admin account"
 				  #   redirect url('/auth/new', false)
 				  # end
 				end
diff --git a/lib/rack_warden/views/rw_layout.html.erb b/lib/rack_warden/views/rw_layout.html.erb
index f2897be..f9111fe 100644
--- a/lib/rack_warden/views/rw_layout.html.erb
+++ b/lib/rack_warden/views/rw_layout.html.erb
@@ -22,17 +22,18 @@
 			<a href="<%=url('/auth/logout', false)%>">Log Out</a>
 		</p>
 
-    <% if flash(:rwarden)[:success] %>
-	    <div style="color:green;">
-	      <%= flash(:rwarden)[:success] %>
-	    </div>
-    <% end %>
-
-    <% if flash(:rwarden)[:error] %>
-	    <div style="color:red;">
-	      <%= flash(:rwarden)[:error] %>
-	    </div>
-    <% end %>
+		<% if flash.rwarden %>
+	    <% if flash.rwarden[:success] %>
+		    <div style="color:green;">
+		      <%= flash.rwarden[:success] %>
+		    </div>
+	    <% end %>
+	    <% if flash.rwarden[:error] %>
+		    <div style="color:red;">
+		      <%= flash.rwarden[:error] %>
+		    </div>
+	    <% end %>
+	   <% end %>
 
 		<div id="content">
 	    <%= yield if block_given? %>
diff --git a/rack_warden.gemspec b/rack_warden.gemspec
index 47e2e96..b484c72 100644
--- a/rack_warden.gemspec
+++ b/rack_warden.gemspec
@@ -19,11 +19,13 @@ Gem::Specification.new do |spec|
   spec.require_paths = ["lib"]
     
   spec.add_dependency "sinatra"
-  spec.add_dependency "sinatra-flash"
+  #spec.add_dependency "sinatra-flash"
+  spec.add_dependency "rack-flash3"
   spec.add_dependency "bcrypt"
   spec.add_dependency "data_mapper"
   spec.add_dependency "dm-sqlite-adapter"
   spec.add_dependency "warden"
+  
 
   spec.add_development_dependency "bundler"
   spec.add_development_dependency "rake"
